src/sdk/pynni/nni/compression/torch/builtin_pruners.py

```diff
@@ -206,25 +259,141 @@ def _get_distance_sum(self, weight, in_idx, out_idx):
+    weights sum in the convolution layers to achieve a preset level of network sparsity.
+    Hao Li, Asim Kadav, Igor Durdanovic, Hanan Samet and Hans Peter Graf,
+    "PRUNING FILTERS FOR EFFICIENT CONVNETS", 2017 ICLR
+    https://arxiv.org/abs/1608.08710
+    """
+
+    def __init__(self, model, config_list):
+        """
+        Parameters
+        ----------
+        model : torch.nn.module
+            Model to be pruned
+        config_list : list
+            List on pruning configs
+        """
+
+        super().__init__(model, config_list)
+        self.if_init_list = {}
+
+    def calc_mask(self, layer, config):
+        """
+        Calculate the mask of given layer.
+        Filters with the smallest sum of its absolute kernel weights are masked.
+        Parameters
+        ----------
+        layer : LayerInfo
+            the layer to instrument the compression operation
+        config : dict
+            layer's pruning config
+        Returns
+        -------
+        torch.Tensor
+            mask of the layer's weight
+        """
+
+        weight = layer.module.weight.data
+        op_name = layer.name
+        op_type = layer.type
+        assert op_type == 'Conv2d', 'FilterPruner only supports 2d convolution layer pruning'
+        if self.if_init_list.get(op_name, True):
+            kernels = weight.shape[0]
+            w_abs = weight.abs()
+            k = int(kernels * config['sparsity'])
+            if k == 0:
+                return torch.ones(weight.shape).type_as(weight)
+            w_abs_structured = w_abs.view(kernels, -1).sum(dim=1)
+            threshold = torch.topk(w_abs_structured.view(-1), k, largest=False).values.max()
+            mask = torch.gt(w_abs_structured, threshold)[:, None, None, None].expand_as(weight).type_as(weight)
+            self.mask_dict.update({op_name: mask})
+            self.if_init_list.update({op_name: False})
+        else:
+            mask = self.mask_dict[op_name]
+        return mask
+
+
+class SlimPruner(Pruner):
+    """
+    A structured pruning algorithm that prunes channels by pruning the weights of BN layers.
+    Zhuang Liu, Jianguo Li, Zhiqiang Shen, Gao Huang, Shoumeng Yan and Changshui Zhang
+    "Learning Efficient Convolutional Networks through Network Slimming", 2017 ICCV
+    https://arxiv.org/pdf/1708.06519.pdf
+    """
+
+    def __init__(self, model, config_list):
+        """
+        Parameters
+        ----------
+        config_list : list
+            List of pruning configs
+        """
+
+        super().__init__(model, config_list)
+        self.if_init_list = {}
+        weight_list = []
+        if len(config_list) > 1:
+            logger.warning('Slim pruner only supports 1 configuration')
+        config = config_list[0]
+        for (layer, config) in self.detect_modules_to_compress():
+            assert layer.type == 'BatchNorm2d', 'SlimPruner only supports 2d batch normalization layer pruning'
```

Comment:
Looks like the slimming algo is not used to prune BatchNorm2d,  it uses Batch norm layer to prune convolutional layers as my understanding,  would you double check this?

Reply:
Yes, in the SlimPruner we prune the channels of convolution layers by masking the corresponding scaling factors in the BN layers.

Reply:
So there is no mask for convolutional layers?   Each conv layer user want to prune should be followed by a batch norm layer?

Reply:
Yes, this algorithm is designed for this situation, as we often use convolution layers as a `conv-bn-relu` combination.

Reply:
got it, thanks!